### PR TITLE
issue14,15,16-Features for view page

### LINKED ISSUE #14: #14add filter function on view page

### LINKED ISSUE #15: #15add sort function on view page

### LINKED ISSUE #16: #16add export report function on view page

### PR DESCRIPTION
**Pull request by yylocky**
# Change Summary 📝

### Features implemented on `view_all_workallocation_page`:

- **Role-Based Access Control:**
  * `HOS` and `Admin` can view **all** works.
  * `HOD` can view **only** their department's works.
  * `Staff` can view **only** their individual work.

- **Department Display Logic:**
  * When related to leaves, the department displayed is the user's department.
  * Otherwise, the work's assigned department is displayed.

- **Table Enhancements:**
  * **Filter:** Added functionality to filter each column.
  * **Sorting:** Columns can be sorted alphabetically or numerically.
  * **Export:** Users can export the table data as a CSV report.

- **Possible Improvement:**
  * Consider introducing an "Other" department, given that `HOS` and `Admin` actually don't belong to any PMC department, currently for simplicity they are associated with the Physics department (`dept id=1`) .

---

## Change Form 📋

- [x] The pull request title references the issue number.
- [x] I have conducted a self-review of my own code.
- [x] I have commented on my code, particularly in complex areas.
- [x] I have updated any relevant documentation.
- [x] Tests (if applicable) have been added or updated.

---

## Background Context 🖼️

- **Since "view_workload" pagename has been used when @kylelhk developing issue25 for filtered workallocation result, the original view_workload has been changed to view_all_workload, thus navbar, routes and related files have been adjusted accordingly.**

- **A more comprehensive testing dataset has been created and uploaded to share with members for testing. refer to the login credential as below:**
Screen Shot 2023-09-19 at 04 05 02

- **Test been conducted:**
 1.  Login as hos,hod,admin,staff respectively, only authorised data is displayed.
 2.  If work category=LEAVE, the department shows the user's department, otherwise task's deparment
 3.  filter function works as expected
 4.  sort function works as expected
 5.  export csv file successfully

---

## Reviewers 🕵️

(Optional) Suggest reviewers that might be interested in this PR.

### CODE REVIEW

**Review by kylelhk — COMMENTED** [review]
Tests conducted and the results are documented as follows:

1.	Logged in as users in different roles and departments to check workload entries in the View Workload page.  All entries rendered on different user’s page match the Role-Based Access Control as described in the Change Summary.

2.	Logged in as users having workload with Leave types work_id and other work_id respectively and got the user’s department and work's assigned department displayed correctly.

3.	Filter and sorting functions work as expected.

4.	The issue below is noted from the exported CSV, which requires a fix on the CSV generation logic:
 
image

Other review comments:

1.	Agree that an "Other" department should be added for HoS and Admin roles.  This can be discussed among all members during any upcoming team meetings – if no potential conflicts / impacts on the existing functions and logics, we may consider this as an enhancement for Deliverable 3.

2.	Per our discussion earlier today, agreed to maintain view_workload.html (for the redirection functions in the Edit Workload and Comment History pages) and view_all_workload.html (for direct access to View Workload page) separately given the relevant limitation in routes.py setting.

3.	No conflict or contradictory element is noted from the code.  This branch is expected to function harmoniously with the main branch.

Approval will be granted upon fixing of the CSV generation.  Well done!

File: app/templates/view_all_workload.html [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+  // Generate CSV content from the rows
+  function generateCSVContent(rows) {
+    const headers = Array.from(
+      document.querySelectorAll(".workload-table th")
+    ).map((header) => header.textContent.trim());
+    const rowsContent = rows.map((row) =>
+      Array.from(row.querySelectorAll("td"))
+        .map((cell) => cell.textContent.trim())
+        .join(",")
+    );
```
Comment (kylelhk) [inline app/templates/view_all_workload.html:255]: The exported CSV is not in order - requires a fix on the generation logic.

**Review by kylelhk — APPROVED** [review]
CSV generation bug fixed per our discussion earlier today.

File: app/templates/view_all_workload.html (hunk 1 above)
Comment (yylocky) [reply app/templates/view_all_workload.html:255]: Thanks Kyle for spotting this 👍🏻

**Pull request merged**